### ISSUE TITLE
Infer annotations for receiver of instanceof checks

### ISSUE THREAD

**Issue by vlsi** [typetools/checker-framework#3616]
Use case:

```java
interface Base {
    @Nullable Object get();
}

class DefaultImpl implements Base {
    /* non-nullable */ Object get() { return new Object(); }
}
```

Usage
```java
void test(Base obj) {
    if (obj instanceof DefaultImpl) {
        obj.get().hashCode(); // <-- nullness verification should pass
    }
}
```

The logic is `DefaultImpl` is known to return non-null object, so the checker should infer that `obj.get()` is non-nullable.

---

Here's a similar case in Apache Calcite codebase: https://github.com/apache/calcite/blob/88d18185e6177c9df587bdd23dd4049f59adc2e4/core/src/main/java/org/apache/calcite/sql/SqlDialect.java#L774-L776

interface: https://github.com/apache/calcite/blob/88d18185e6177c9df587bdd23dd4049f59adc2e4/core/src/main/java/org/apache/calcite/rel/type/RelDataType.java#L192

implementation: https://github.com/apache/calcite/blob/88d18185e6177c9df587bdd23dd4049f59adc2e4/core/src/main/java/org/apache/calcite/sql/type/AbstractSqlType.java#L61-L62 and https://github.com/apache/calcite/blob/88d18185e6177c9df587bdd23dd4049f59adc2e4/core/src/main/java/org/apache/calcite/sql/type/BasicSqlType.java#L34

**Comment by smillst** [comment]
This was fixed by 22d0286 .
